### PR TITLE
fix: deduplicate repeated chapter names in output

### PR DESCRIPTION
**Pull request by eton-s**
Fixes silent merging of highlights between chapters with duplicate names. Part of a set of changes I made to the plugin with Claude Code to improve my Obsidian workflow.

---

## Summary

Books that reuse the same chapter title across multiple parts (e.g. a book structured as "Part One / Chapter 1, Part Two / Chapter 1") previously had all highlights silently merged under a single `## Chapter 1` heading.

This fix groups highlights by `contentId` rather than by chapter title. The new `buildChapterList` method returns an array of `[title, bookmarks]` tuples instead of a Map, so duplicate chapter titles coexist naturally — each distinct content section gets its own heading using the original chapter name, with no renaming or suffixing.

## Changes

- src/database/Highlight.ts — add `buildChapterList()`: groups highlights by `contentId`, returns `[chapter, Bookmark[]][]` with original titles preserved
- src/template/template.ts — `applyTemplateTransformations` now accepts `[chapter, Bookmark[]][]` directly (it was already converting the Map to this shape internally)
- src/modal/ExtractHighlightsModal.ts — rewrite `fetchHighlights` / `writeBooks` to group highlights per book and call `buildChapterList` instead of the previous `convertToMap` path
- src/template/template.test.ts — update test data to use arrays; add test that calls `buildChapterList` with duplicate-titled `Highlight` objects and verifies both `## Chapter 1` headings appear in the rendered output

## Behaviour change

Books with unique chapter titles are unaffected. Books with repeated chapter titles now get separate headings with the same name (e.g. two `## Chapter 1` headings) rather than having highlights silently merged. Chapter order is maintained by the sort in `getAllHighlight`.

## Dead Code

`convertToMap` and `createEmptyContentMap` in `Highlight.ts` are no longer called in the application flow — the modal now groups by book title directly and uses `buildChapterList` per book. Both methods still exist and their tests pass.

---

> **Note:** The implementation in this PR was developed with AI assistance
> (Claude). The logic has been manually reviewed and tested against a real
> Kobo device, but please flag anything that looks wrong.

### CODE REVIEW

File: src/template/template.test.ts [hunk 1]
```diff
@@ -209,4 +209,47 @@ test2
 			chai.expect(normalize(content)).equal(normalize(t[1]));
 		});
 	}
+
+	it("applyTemplateTransformations with dedup chapters", async function () {
+		// Simulates a book where two distinct content entries share the same
+		// chapter title (e.g. "Chapter 1" appears in both Part One and Part Two).
+		// buildChapterMapWithDedup should give them unique keys so highlights
+		// are never silently merged.
+		const dedupChapters = new Map<string, Bookmark[]>([
```
Comment (OGKevin) [inline src/template/template.test.ts]: why build the dedup map manually instead of using `buildChapterMapWithDedup`?

File: src/database/Highlight.ts [hunk 2]
```diff
@@ -177,4 +177,34 @@ export class HighlightService {
...
+			if (key === undefined) {
+				const count = (titleCount.get(title) ?? 0) + 1;
+				titleCount.set(title, count);
+				key = count === 1 ? title : `${title} (${count})`;
+				contentIdToKey.set(contentId, key);
+				chapterMap.set(key, []);
+			}
+
+			chapterMap.get(key)!.push(h.bookmark);
+		}
```
Comment (OGKevin) [inline src/database/Highlight.ts]: There is also a collision here, if the title `key` -> `Chapter 1 (2)` is an actual chapter within the book? 

```
Given highlights ordered: [{title:"A", contentId:"1"}, {title:"A", contentId:"2"}, {title:"A (2)", contentId:"3"}]

contentId "1" → titleCount("A")=1 → key="A"
contentId "2" → titleCount("A")=2 → key="A (2)"
contentId "3" → titleCount("A (2)")=1 → key="A (2)" → chapterMap.set("A (2)", []) overwrites the array from step 2
Result: all bookmarks for contentId "2" are lost.
```

So key must be something unique.

**Review by OGKevin — CHANGES_REQUESTED** [review]

File: src/database/Highlight.ts (hunk 2 above)
Comment (eton-s) [reply src/database/Highlight.ts]: Thanks for catching this. Any suffix scheme has the same problem where a generated key might be a chapter title already in the book, so I've reworked the approach to avoid suffixes and allow for duplicate chapter titles in the note itself.

`buildChapterMapWithDedup` is now `buildChapterList`, and instead of returning a Map, it returns an array of `[title, bookmarks]` tuples grouped by `contentId`. Each chapter keeps its original title verbatim, and now that #519 is merged and chapter order is correct, the identity of each duplicate chapter is clear from the order they appear in. Changing from Map to array also required a small change to `template.ts`, but `chapters` was already being converted to an array anyways, so no change to custom template behavior.

This also addresses your other comment about the test, which now calls `buildChapterList` directly. Since the expected behavior has changed (no more suffixes), the assert is now that both entries have title `"Chapter 1"`.

One note: `convertToMap` and `createEmptyContentMap` are no longer called in the modal flow since `buildChapterList` is used instead. I think they are dead code now, can be removed in this PR or left alone if you prefer.

File: src/template/template.test.ts (hunk 1 above)
Comment (eton-s) [reply src/template/template.test.ts]: fixed, see my other comment for more details
